{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17daeff7",
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import os\n",
    "from collections import defaultdict\n",
    "import numpy as np\n",
    "import scipy.stats as stats \n",
    "from scipy.stats import wilcoxon\n",
    "\n",
    "import sys\n",
    "sys.path.append('/opt/miniconda3/lib/python3.8/site-packages')\n",
    "from matplotlib_venn import venn3\n",
    "\n",
    "pd.set_option('display.max_columns', None)\n",
    "pd.set_option('display.max_rows', 110)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28f90bb2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "0ed87ab7",
   "metadata": {},
   "source": [
    "# Change-o"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a38b1059",
   "metadata": {},
   "source": [
    "To process 10X V(D)J data, a combination of AssignGenes.py and MakeDb.py can be used to generate a TSV file compliant with the AIRR Community Rearrangement schema that incorporates annotation information provided by the Cell Ranger pipeline. The --10x filtered_contig_annotations.csv specifies the path of the contig annotations file generated by cellranger vdj, which can be found in the outs directory."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b5cf188",
   "metadata": {},
   "outputs": [],
   "source": [
    "#sample 1\n",
    "!AssignGenes.py igblast -s ./changeo_input/filtered_contig1.fasta -b ~/share/igblast \\\n",
    "    --organism human --loci ig --format blast --exec /Users/elinorwing/Documents/Python/ncbi-igblast-1.18.0/bin/igblastn "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d2c492e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#sample 2\n",
    "!AssignGenes.py igblast -s ./changeo_input/filtered_contig2.fasta -b ~/share/igblast \\\n",
    "   --organism human --loci ig --format blast  --exec /Users/elinorwing/Documents/Python/ncbi-igblast-1.18.0/bin/igblastn "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a5339110",
   "metadata": {},
   "outputs": [],
   "source": [
    "#sample 3\n",
    "!AssignGenes.py igblast -s ./changeo_input/filtered_contig3.fasta -b ~/share/igblast \\\n",
    "   --organism human --loci ig --format blast --exec /Users/elinorwing/Documents/Python/ncbi-igblast-1.18.0/bin/igblastn "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30e7ff20",
   "metadata": {},
   "outputs": [],
   "source": [
    "!MakeDb.py igblast --log logfil.txt -i ./changeo_input/filtered_contig1_igblast.fmt7 -s ./changeo_input/filtered_contig1.fasta \\\n",
    "   -r /Users/elinorwing/share/germlines/imgt/human/vdj/imgt_human_*.fasta --10x ./changeo_input/filtered_contig1_annotations.csv --failed --extended "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dc2a8d68",
   "metadata": {},
   "outputs": [],
   "source": [
    "!MakeDb.py igblast --log logfil.txt -i ./changeo_input/filtered_contig2_igblast.fmt7 -s ./changeo_input/filtered_contig2.fasta \\\n",
    "   -r /Users/elinorwing/share/germlines/imgt/human/vdj/imgt_human_*.fasta --10x ./changeo_input/filtered_contig2_annotations.csv --failed --extended"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5115a4a",
   "metadata": {},
   "outputs": [],
   "source": [
    "!MakeDb.py igblast --log logfile.txt -i ./changeo_input/filtered_contig3_igblast.fmt7 -s ./changeo_input/filtered_contig3.fasta \\\n",
    "   -r /Users/elinorwing/share/germlines/imgt/human/vdj/imgt_human_*.fasta --10x ./changeo_input/filtered_contig3_annotations.csv --failed --extended"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f0f74d7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "c33e0e66",
   "metadata": {},
   "source": [
    "# Import Change-o output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e45fa381",
   "metadata": {},
   "outputs": [],
   "source": [
    "source_folder = \"/Users/elinorwing/Documents/Edinburgh/Year 2/scRNA_seq/analysis/bcr/NEW BCR/clones/input2\"\n",
    "\n",
    "file_list = [x for x in os.listdir(f'{source_folder}') if x.endswith('.tsv')]\n",
    "\n",
    "all_df = pd.DataFrame()\n",
    "\n",
    "for file in file_list:    \n",
    "    file_df = pd.read_csv(f'{source_folder}/{file}', sep='\\t')\n",
    "    file_df['sample'] =  file.split(\"_\")[0]\n",
    "    all_df = all_df.append(file_df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "27687bb9",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a5976a97",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_df.groupby('sample')['cell_id'].agg(total = \"nunique\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec2ab8ba",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "58d08547",
   "metadata": {},
   "source": [
    "# Get only cells with one single IGH and IGK/IGL chain and get SHM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca6a549d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#AAACCTGAGCTGTTCA-1_contig_1 split by the first '_', 0 is AAA...-1, 1 is contig_1\n",
    "all_df[\"contig\"] = all_df[\"sequence_id\"].str.split(\"_\", 1).str[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5d98a70a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#if the locus column contains 'IGH' add 'Heavy to the chain column, otherwise add 'Light'\n",
    "all_df[\"chain\"] = np.where(all_df['locus'].str.contains('IGH'), \"Heavy\" , \"Light\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1927fc0",
   "metadata": {},
   "outputs": [],
   "source": [
    "chain_counts = all_df.groupby(['sample', 'chain'])['cell_id'].agg(number_chains = \"value_counts\")\n",
    "chain_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73d6973f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#new column with proper gene names - IGHV3-15*07 becomes IGHV3-15\n",
    "all_df[\"v_gene\"] = all_df[\"v_call\"].str.split(\"*\").str[0]\n",
    "all_df[\"d_gene\"] = all_df[\"d_call\"].str.split(\"*\").str[0]\n",
    "all_df[\"j_gene\"] = all_df[\"j_call\"].str.split(\"*\").str[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "070cda27",
   "metadata": {},
   "outputs": [],
   "source": [
    "#new column with variable chain family name - IGHV3-15*07 becomes IGHV3\n",
    "all_df[\"v_family\"] = all_df[\"v_call\"].str.split(\"-\").str[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d94d34d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#identify duplicates where cell_id, sample, and chain are the same\n",
    "multiple_chains = all_df[all_df.duplicated(subset = [\"cell_id\", \"sample\", \"chain\"], keep = False)].copy()\n",
    "\n",
    "multiple_chains"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8717939c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#remove the duplicates\n",
    "all_df = all_df.drop_duplicates(subset = [\"cell_id\", \"sample\", \"chain\"], keep = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68552c93",
   "metadata": {},
   "outputs": [],
   "source": [
    "#remove rows where there is no sequence alignment or germline alignment\n",
    "all_df_shm = all_df.dropna(subset = [\"sequence_alignment\", \"germline_alignment\"]).copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6cebb4a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_mut_count_v(sequence, germline):\n",
    "    mutations = 0\n",
    "    for i, base in enumerate(sequence[:312]):\n",
    "        if base != \".\" and base != \"-\" and base != \"N\" and germline[i] != \"N\" and germline[i] != \"-\" and germline[i] !=  \".\" :\n",
    "            if base != germline[i]:\n",
    "                mutations += 1\n",
    "    return mutations\n",
    "  \n",
    "def get_mut_freq_v(sequence, germline):\n",
    "    mutations = 0\n",
    "    length = 0\n",
    "    for i, base in enumerate(sequence[:312]):\n",
    "        if base != \".\" and base != \"-\" and base != \"N\" and germline[i] != \"N\" and germline[i] != \"-\" and germline[i] !=  \".\" :\n",
    "            length += 1\n",
    "            if base != germline[i]:\n",
    "                mutations += 1\n",
    "    if length == 0:\n",
    "        return np.nan\n",
    "    else:\n",
    "        return mutations/length"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c48ed3f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#determine the mutation count and frequency in the variable regions using Catherines functions\n",
    "all_df_shm['mut_count_v'] = all_df_shm.apply(lambda row : get_mut_count_v(row['sequence_alignment'],\n",
    "                 row['germline_alignment']), axis = 1)\n",
    "\n",
    "all_df_shm['mut_freq_v'] = all_df_shm.apply(lambda row : get_mut_freq_v(row['sequence_alignment'],\n",
    "                 row['germline_alignment']), axis = 1)\n",
    "\n",
    "\n",
    "all_df_shm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "815ed1e4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "5488c791",
   "metadata": {},
   "source": [
    "# Combine heavy and light annotations into single row per cell"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a396116",
   "metadata": {},
   "outputs": [],
   "source": [
    "#all_df_igh contains all of all_df where the locus is IGH (heavy)\n",
    "#all_df_ighlk contains the info where the locus is not IGH - IGL or IGK (light)\n",
    "all_df_igh = all_df_shm.query(\"locus == 'IGH'\")\n",
    "all_df_iglk = all_df_shm.query(\"locus != 'IGH'\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a5cb7e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#add heavy_ to the beginning of all column names in heavy df and light_ to all columns in light df\n",
    "all_df_igh = all_df_igh.add_prefix('heavy_')\n",
    "all_df_iglk = all_df_iglk.add_prefix('light_')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bbe8568a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#recreate column names containing sample and cell id info for merging\n",
    "all_df_igh[\"sample\"] = all_df_igh[\"heavy_sample\"]\n",
    "all_df_iglk[\"sample\"] = all_df_iglk[\"light_sample\"]\n",
    "\n",
    "\n",
    "all_df_igh[\"cell_id\"] = all_df_igh[\"heavy_cell_id\"]\n",
    "all_df_iglk[\"cell_id\"] = all_df_iglk[\"light_cell_id\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6a5d60a",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_df_igh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82d91303",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_df_iglk"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "75ff7cba",
   "metadata": {},
   "outputs": [],
   "source": [
    "#combine the two dfs\n",
    "all_df_both = pd.merge(all_df_igh, all_df_iglk, how=\"outer\", on=[\"cell_id\", \"sample\"])\n",
    "all_df_both"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cdf14976",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_df_both_paired = pd.merge(all_df_igh, all_df_iglk, how=\"inner\", on=[\"cell_id\", \"sample\"])\n",
    "all_df_both_with_heavy = all_df_both.dropna(subset = [\"heavy_sequence_id\"])\n",
    "\n",
    "all_df_both_paired"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "32abbb3b",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_df_both_with_heavy.to_csv(\"all_df_both_with_heavy_new_2.csv\")\n",
    "all_df_both_paired.to_csv(\"all_df_both_paired_new_2.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f9536e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "list(all_df_both_paired.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6bceb51c",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_df_both_paired = pd.read_csv(\"all_df_both_paired_new.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f408f87",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "aa28e16d",
   "metadata": {},
   "source": [
    "# Tidy columns and sort ordering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29bf391a",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_df_both_paired = all_df_both_paired.sort_values(by = \"sample\")    \n",
    "individ_order = list(all_df_both_paired[\"sample\"].unique())\n",
    "individ_order"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4bba3bb",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "cc2fffcd",
   "metadata": {},
   "source": [
    "# Number of heavy or light chains per cell"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b4d9840",
   "metadata": {},
   "outputs": [],
   "source": [
    "chain_counts.groupby(['sample', 'chain'])[\"number_chains\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45da2d2c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "6e4ef185",
   "metadata": {},
   "source": [
    "# Get percentage of all cells that have a single paired receptor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "91ca7e0f",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_grpd = pd.DataFrame(all_df.groupby('sample')['cell_id'].agg(total_cell_count = \"nunique\"))\n",
    "\n",
    "single_pair_total = pd.DataFrame(all_df_both_paired.groupby('sample')['cell_id'].agg(total_single = \"nunique\").copy())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28f317ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_grpd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5e3913a",
   "metadata": {},
   "outputs": [],
   "source": [
    "single_pair_total[\"total_cell_count\"] = [3567, 4744, 5793]\n",
    "single_pair_total"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec0da089",
   "metadata": {},
   "outputs": [],
   "source": [
    "single_pair_total[\"percent_single_pair_total\"] = single_pair_total[\"total_single\"]/single_pair_total[\"total_cell_count\"] * 100\n",
    "single_pair_total"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c870fe07",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "cc338997",
   "metadata": {},
   "source": [
    "# Output for Shazam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "99ef51b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "for_shazam = all_df_shm.drop_duplicates(subset = [\"cell_id\", \"sample\", \"chain\"], keep = False)\n",
    "for_shazam = for_shazam.dropna(subset = [\"junction\"])\n",
    "for_shazam[\"cell_id_sample\"] = for_shazam[\"cell_id\"] + for_shazam[\"sample\"]\n",
    "\n",
    "\n",
    "for_shazam.to_csv(\"all_df_for_shazam.csv\", index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "082c12e2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "bb1c6ad8",
   "metadata": {},
   "source": [
    "# Plot isotype usage per sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17c920b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "isotype_use = all_df_both_paired.copy()\n",
    "isotype_use[\"count\"] = 1\n",
    "isotype_use = isotype_use.sort_values(by = \"sample\")    \n",
    "individ_order = list(isotype_use[\"sample\"].unique())\n",
    "\n",
    "isotype_use = isotype_use[[\"sample\", \"count\", \"heavy_c_call\"]]\n",
    "\n",
    "isotype_use = isotype_use.groupby(['sample', 'heavy_c_call']).agg({'count': 'sum'})\n",
    "isotype_use = isotype_use.groupby(level=0).apply(lambda x: x / float(x.sum())).reset_index()\n",
    "\n",
    "isotype_use = isotype_use.pivot(index = \"sample\", values = \"count\", columns = \"heavy_c_call\").reset_index()\n",
    "\n",
    "isotype_use.individ_type=pd.Categorical(isotype_use.sample,categories=individ_order)\n",
    "isotype_use=isotype_use.sort_values('sample')\n",
    "\n",
    "isotype_use"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c6b5b2ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "f = plt.figure()\n",
    "\n",
    "isotype_use.plot(\n",
    "    x = 'sample',\n",
    "    kind = 'bar',\n",
    "    stacked = True,\n",
    "    width = 0.9,\n",
    "    figsize=(8,4),\n",
    "    legend = True,\n",
    "    colormap = \"Set2\",\n",
    "    ax = f.gca());\n",
    "plt.legend(loc='center left', bbox_to_anchor=(1.0, 0.5))\n",
    "plt.xlabel(\"Sample\")\n",
    "plt.xticks(rotation=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "80fad394",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "45165a72",
   "metadata": {},
   "source": [
    "# Somatic Hypermutation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82ab4285",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_df_both_paired"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "271db343",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_df_both_paired[\"heavy_mut_freq_v\"] = all_df_both_paired[\"heavy_mut_freq_v\"].astype(int)\n",
    "all_df_both_paired[\"light_mut_freq_v\"] = all_df_both_paired[\"light_mut_freq_v\"].astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bcb44944",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_df_both_paired_shm = all_df_both_paired.dropna(subset = [\"heavy_c_call\"]).copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "83bc6175",
   "metadata": {},
   "outputs": [],
   "source": [
    "igg_shm_df = all_df_both_paired_shm[all_df_both_paired_shm[\"heavy_c_call\"].str.contains(\"IGHG\")]\n",
    "iga_shm_df = all_df_both_paired_shm[all_df_both_paired_shm[\"heavy_c_call\"].str.contains(\"IGHA\")]\n",
    "igm_shm_df = all_df_both_paired_shm[all_df_both_paired_shm[\"heavy_c_call\"].str.contains(\"IGHM\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d81362d7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e774689",
   "metadata": {},
   "outputs": [],
   "source": [
    "g = sns.relplot(data = igg_shm_df, x = \"heavy_mut_freq_v\", y = \"light_mut_freq_v\", alpha = 0.10, col = \"heavy_sample\", col_wrap = 2)\n",
    "g.set_xlabels('Mutation Frequency IgH V Gene');\n",
    "g.set_ylabels('Mutation Frequency IgK/L V Gene');\n",
    "g.set_titles(col_template = '{col_name}');"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ddd683a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "no_igd = all_df_both_paired.query(\"heavy_c_call!='IGHD'\")\n",
    "\n",
    "sns.displot(\n",
    "    no_igd, x=\"heavy_mut_freq_v\", hue=\"heavy_sample\", col='heavy_c_call', col_wrap=4,\n",
    "    height=4, facet_kws=dict(margin_titles=True), common_norm = False, kind='kde', \n",
    "    fill=True, palette=sns.color_palette('bright')[:3]\n",
    ");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "940d3942",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "9855687d",
   "metadata": {},
   "source": [
    "# Overlap unique heavy chain sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e934acd",
   "metadata": {},
   "outputs": [],
   "source": [
    "set1 = set(all_df_both_paired_shm.query(\"heavy_sample == 1\")['heavy_cdr3'])\n",
    "set2 = set(all_df_both_paired_shm.query(\"heavy_sample == 2\")['heavy_cdr3'])\n",
    "set3 = set(all_df_both_paired_shm.query(\"heavy_sample == 3\")['heavy_cdr3'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b138f449",
   "metadata": {},
   "outputs": [],
   "source": [
    "venn3([set1, set2, set3], ('Sample1', 'Sample2', 'Sample3'))\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c56c483",
   "metadata": {},
   "outputs": [],
   "source": [
    "hj1 = set(all_df_both_paired_shm.query(\"heavy_sample == 1\")['heavy_junction_aa'])\n",
    "hj2 = set(all_df_both_paired_shm.query(\"heavy_sample == 2\")['heavy_junction_aa'])\n",
    "hj3 = set(all_df_both_paired_shm.query(\"heavy_sample == 3\")['heavy_junction_aa'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1621182f",
   "metadata": {},
   "outputs": [],
   "source": [
    "venn3([hj1, hj2, hj3], ('Sample1', 'Sample2', 'Sample3'))\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0522a222",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "af843f29",
   "metadata": {},
   "source": [
    "# CDR3 Lengths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9146041b",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_df_both_paired = pd.read_csv(\"all_df_both_paired_new.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cd257d7d",
   "metadata": {},
   "outputs": [],
   "source": [
    "mean_cdr3_lengths = all_df_both_paired.copy()\n",
    "\n",
    "mean_cdr3_lengths = mean_cdr3_lengths.groupby([\"sample\"])[\"heavy_junction_length\"].mean().reset_index()\n",
    "\n",
    "mean_cdr3_lengths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ffa67da",
   "metadata": {},
   "outputs": [],
   "source": [
    "mean_cdr3_lengths_iga = all_df_both_paired.dropna(subset =[\"heavy_c_call\"]).copy()\n",
    "\n",
    "mean_cdr3_lengths_iga = mean_cdr3_lengths_iga[mean_cdr3_lengths_iga[\"heavy_c_call\"].str.contains(\"IGHA\")].copy()\n",
    "\n",
    "mean_cdr3_lengths_iga = mean_cdr3_lengths_iga.groupby([\"sample\"])[\"heavy_junction_length\"].mean().reset_index()\n",
    "\n",
    "mean_cdr3_lengths_iga"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5431af45",
   "metadata": {},
   "outputs": [],
   "source": [
    "mean_cdr3_lengths_igg = all_df_both_paired.dropna(subset =[\"heavy_c_call\"]).copy()\n",
    "\n",
    "mean_cdr3_lengths_igg = mean_cdr3_lengths_igg[mean_cdr3_lengths_igg[\"heavy_c_call\"].str.contains(\"IGHG\")].copy()\n",
    "\n",
    "mean_cdr3_lengths_igg = mean_cdr3_lengths_igg.groupby([\"sample\"])[\"heavy_junction_length\"].mean().reset_index()\n",
    "\n",
    "mean_cdr3_lengths_igg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed5d91d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "mean_cdr3_lengths_igm = all_df_both_paired.dropna(subset =[\"heavy_c_call\"]).copy()\n",
    "\n",
    "mean_cdr3_lengths_igm = mean_cdr3_lengths_igm[mean_cdr3_lengths_igm[\"heavy_c_call\"].str.contains(\"IGHM\")].copy()\n",
    "\n",
    "mean_cdr3_lengths_igm = mean_cdr3_lengths_igm.groupby([\"sample\"])[\"heavy_junction_length\"].mean().reset_index()\n",
    "\n",
    "mean_cdr3_lengths_igm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "79716216",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "6d93f1a7",
   "metadata": {},
   "source": [
    "# Gini Coefficient"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14b8dbf6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_gini(array):\n",
    "    \"\"\"Calculate the Gini coefficient of a numpy array.\"\"\"\n",
    "    # based on bottom eq:\n",
    "    # http://www.statsdirect.com/help/generatedimages/equations/equation154.svg\n",
    "    # from:\n",
    "    # http://www.statsdirect.com/help/default.htm#nonparametric_methods/gini.htm\n",
    "    # All values are treated equally, arrays must be 1d:\n",
    "    array = array.flatten()\n",
    "    if np.amin(array) < 0:\n",
    "        # Values cannot be negative:\n",
    "        array -= np.amin(array)\n",
    "    # Values cannot be 0:\n",
    "    array += 0.0000001\n",
    "    # Values must be sorted:\n",
    "    array = np.sort(array)\n",
    "    # Index per array element:\n",
    "    index = np.arange(1,array.shape[0]+1)\n",
    "    # Number of array elements:\n",
    "    n = array.shape[0]\n",
    "    # Gini coefficient:\n",
    "    return ((np.sum((2 * index - n  - 1) * array)) / (n * np.sum(array)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "91651278",
   "metadata": {},
   "outputs": [],
   "source": [
    "unique_cdrs = unique_cdrs.sort_values(by = \"sample\")\n",
    "\n",
    "plot_gini_dict = {}\n",
    "\n",
    "for i in unique_cdrs[\"sample\"].unique():\n",
    "    sample_df = unique_cdrs[unique_cdrs[\"sample\"] == i]\n",
    "    temp_array = sample_df.loc[:,'cell_prop'].values\n",
    "    temp_array = temp_array[~np.isnan(temp_array)]\n",
    "    gini = calculate_gini(temp_array)\n",
    "    plot_gini_dict[i] = gini\n",
    "\n",
    "plot_gini_df = pd.DataFrame.from_dict(plot_gini_dict, orient = \"index\", columns = [\"gini_coeffic\"]).reset_index()\n",
    "plot_gini_df.rename(columns={'index': 'sample'}, inplace = True)\n",
    "\n",
    "\n",
    "with sns.plotting_context(\"notebook\", font_scale = 2):\n",
    "    plt.figure(figsize = (16,8))\n",
    "\n",
    "    g = sns.stripplot(x=\"sample\", y=\"gini_coeffic\", data=plot_gini_df,\n",
    "                linewidth=1, s= 7, color = \"black\")\n",
    "    sns.boxplot(x=\"sample\",\n",
    "            y=\"gini_coeffic\",\n",
    "            data=plot_gini_df,\n",
    "            showfliers = False)\n",
    "    plt.xticks(rotation=45, ha=\"right\")\n",
    "    plt.ylabel(\"Gini Coefficient\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "442b8182",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_gini_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d8de877a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "05e08a7d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ddfd7a13",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92c513ef",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "07e1846f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "794e88c1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ecf4408f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30ba968f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
